File: agent-driven-architecture-main/examples/typescript/src/blocks/artifact/contract.ts
// ── blocks/artifact/contract — the work product's TRANSPORT (G16) ──────────
// The artifact was built by PERFORMED EFFECTS, which replay stubs — so 2.2's
// "the folded, replayable result of the session" was false, and a reducer
// change that corrupted artifact content while leaving State byte-identical
// passed every check on offer.
//
// Here the artifact IS State. `recordFinding` folds a line. Delivery is ONE
// irreversible effect at seal time, gated by G6 exactly as 14.3 says session-end
// is — not one effect per line performed as you go.

import type { CommandBase } from "@adr/spine/pure/command";
import type { EffectBase } from "@adr/spine/pure/effect";
import type { ToolResultBase } from "@adr/spine/pure/tool-result";
import { claims } from "@adr/spine/pure/tool-result";
import type { ArtifactLine } from "./slice";

// ── ToolResult cases ────────────────────────────────────────────────────────
export interface RecordFindingResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "recordFinding";
  readonly text: string;
}

export interface RequestSealResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "requestSeal";
}

export interface ConfirmSealResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "confirmSeal";
}

export type ArtifactResult = RecordFindingResult | RequestSealResult | ConfirmSealResult;

// ── Command cases ───────────────────────────────────────────────────────────
export interface RecordFindingCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "recordFinding";
  readonly text: string;
}

export interface RequestSealCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "requestSeal";
}

export interface ConfirmSealCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "confirmSeal";
}

export type ArtifactCommand = RecordFindingCommand | RequestSealCommand | ConfirmSealCommand;

// ── Effect cases ────────────────────────────────────────────────────────────
/** IRREVERSIBLE, and it fires EXACTLY ONCE, at seal. The lines it carries are a
 *  value copied out of committed State, so what was delivered is what replay
 *  re-derives. */
export interface DeliverArtifact extends EffectBase {
  readonly kind: "DeliverArtifact";
  /** IRREVERSIBLE: the work product leaves the system. Its verb `confirmSeal` is
   *  the registry's other `Irreversible` row — the same fact stated at the other
   *  end of the same step (docs/DECISIONS.md:85). */
  readonly effectClass: "Irreversible";
  readonly lines: readonly ArtifactLine[];
}

export type ArtifactEffect = DeliverArtifact;

/** WHICH RESULTS THIS BLOCK'S ARM FOLDS, derived rather than written. The table
 *  is a mapped type over the union above (`spine/pure/tool-result`), so a case
 *  added there and a case claimed here are ONE edit: omit it and the property is
 *  missing, name a tool the union does not declare and the property is excess.
 *  The under-claiming predicate this used to be — green build, stale clause,
 *  fall-through at run time — is not writable in this form. */
export const isArtifactResult = claims<ArtifactResult>({
  recordFinding: true,
  requestSeal: true,
  confirmSeal: true,
});

File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/blocks/artifact/Contract.kt
// ── blocks/artifact/contract — the work product's transport (G16) ──────────
// The artifact used to be built by PERFORMED EFFECTS — which replay stubs — so 2.2's
// "the folded, replayable result of the session" was false, and a reducer change
// that corrupted artifact content while leaving State byte-identical passed every
// check on offer.
//
// Here the artifact IS State: one line per fold arm. Delivery is ONE irreversible
// effect at seal time, gated by G6 exactly as 14.3 says session-end is.

package adr.contract

import adr.spine.pure.Actor
import adr.spine.pure.CommandId
import adr.spine.pure.EffectClass
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName

/**
 * A sealed CLASS extending the sealed CLASS ToolResult: `tool` is passed up the chain,
 * so every variant carries it by construction rather than by re-implementing it.
 */
sealed class ArtifactResult(override val tool: ToolName) : ToolResult(tool) {

    /**
     * FORCED INTO THE TRANSPORT ROOT by Kotlin's sealed rule, not by preference.
     *
     * Every variant of a sealed hierarchy must live in one package AND one MODULE, so
     * this block's transport is authored inside `:spine` (ADR-001 §3) — the compiler
     * says "Extending sealed classes or interfaces from a different module is
     * prohibited". A payload type the transport NAMES therefore cannot stay behind in
     * the block module: `:spine` may not depend on a block at all.
     *
     * It is nested on the block's own sealed root rather than promoted to
     * `adr.spine.pure`, because the kernel is a VENDORABLE tier (gate check C15) and
     * an artifact line is block domain vocabulary — putting it there would be a
     * worse violation of "the spine tier names nothing in your feature code" than
     * anything the DAG fixes. Nested, C2 admits it by the existing name-prefix rule
     * (`ArtifactResult.ArtifactLine` starts with `Artifact`), C15 never sees it, `adr.spine.pure`
     * gains nothing, and no new file and no new package enter the kernel.
     *
     * `by` is the stamped Actor, copied in by the ARM from `sig` — never by the tool.
     * Consumers import the nested name, so blocks/artifact still reads `ArtifactLine`.
     */
    data class ArtifactLine(val at: Timestamp, val by: Actor, val text: String)

    data class RecordFinding(
        override val tool: ToolName,
        val text: String,
    ) : ArtifactResult(tool)

    data class RequestSeal(override val tool: ToolName) : ArtifactResult(tool)

    data class ConfirmSeal(override val tool: ToolName) : ArtifactResult(tool)
}

/**
 * A sealed CLASS extending the sealed CLASS Command: tool/sig/id pass up the chain and
 * every variant carries authorship, permission and identity by construction (G12).
 */
sealed class ArtifactCommand(
    override val tool: ToolName,
    override val sig: Signature,
    override val id: CommandId,
) : Command(tool, sig, id) {
    data class RecordFinding(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val text: String,
    ) : ArtifactCommand(tool, sig, id)

    data class RequestSeal(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
    ) : ArtifactCommand(tool, sig, id)

    data class ConfirmSeal(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
    ) : ArtifactCommand(tool, sig, id)
}

sealed class ArtifactEffect(
    override val at: Timestamp,
    effectClass: EffectClass,
) : Effect(at, effectClass) {
    /** IRREVERSIBLE, and it fires exactly ONCE, at seal time — never once per line: the
     *  work product leaves the system. Its verb `confirmSeal` is the registry's other
     *  Irreversible row — the same fact stated at the other end of the same step. */
    data class DeliverArtifact(
        override val at: Timestamp,
        val lines: List<ArtifactResult.ArtifactLine>,
    ) : ArtifactEffect(at, EffectClass.Irreversible)
}
